components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,243 @@
+ *
+ * @param queue Send queue
+ */
+static void virtio_vsock_notify_free_send(virtqueue_driver_t *queue)
+{
+    void *buf = NULL;
+    uint32_t buf_size = 0, wr_len = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_used_buf(queue, &handle, &wr_len)) {
+        while (camkes_virtqueue_driver_gather_buffer(queue, &handle, &buf, &buf_size, &flag) >= 0) {
+            camkes_virtqueue_buffer_free(queue, buf);
+        }
+    }
+}
+
+/* Should be defined in virtio_vsock_emul.c */
+void WEAK vsock_rx_complete(virtio_emul_t *emul, char *buf, unsigned len)
+{
+    assert(!"should not be here");
+}
+
+/**
+ * Receives data from the source VM using camkes virtqueue, and invokes the emul layer
+ * to transmit the data to the guest VM.
+ *
+ * @param queue Receive queue
+ */
+static void virtio_vsock_notify_recv(virtqueue_device_t *queue)
+{
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_available_buf(queue, &handle)) {
+        char temp_buf[VIRTIO_VSOCK_CAMKES_MTU];
+        size_t len = virtqueue_scattered_available_size(queue, &handle);
+        if (camkes_virtqueue_device_gather_copy_buffer(queue, &handle, (void *)temp_buf, len) < 0) {
+            ZF_LOGW("Dropping data: Can't gather vq buffer.");
+            continue;
+        }
+
+        vsock_rx_complete(virtio_vsock->emul, temp_buf, len);
+        queue->notify();
+    }
+}
+
+/**
+ * Callback function for camkes virtqueue notification. Polls send queues and receive queues
+ * to see if there are data pending.
+ *
+ * @see main.c (struct device_notify) for parameters
+ */
+void handle_vsock_irq(vm_t *vmm)
+{
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].recv_queue && VQ_DEV_POLL(connections[i].recv_queue)) {
+            virtio_vsock_notify_recv(connections[i].recv_queue);
+        }
+        if (connections[i].send_queue && VQ_DRV_POLL(connections[i].send_queue)) {
+            virtio_vsock_notify_free_send(connections[i].send_queue);
+        }
+    }
+}
+
+/**
+ * Callback function for the emul layer. Forwards data to the destination VM using camkes virtqueue.
+ *
+ * @see virtio_pci_vsock.h for function format
+ *
+ * @param cid Dest context ID
+ * @param buffer Data to forward
+ * @param len Length of the data
+ */
+static void tx_forward(int cid, void *buffer, unsigned int len)
+{
+    /* Find the CID */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].cid == cid) {
+            int err = camkes_virtqueue_driver_scatter_send_buffer(connections[i].send_queue, buffer, len);
+            connections[i].send_queue->notify();
+            return;
+        }
+    }
+
+    ZF_LOGE("Invalid CID: %d", cid);
+}
+
+/**
+ * Sets up virtio vsock connections.
+ */
+static void make_virtio_vsock_virtqueues(void)
+{
+    int err;
+    for (int i = 0; i < NUM_PORTS; i++) {
+        virtqueue_driver_t *vq_send;
+        virtqueue_device_t *vq_recv;
+
+        vq_recv = malloc(sizeof(*vq_recv));
```

Comment:
Sometime we use `malloc()` and sometime `calloc()`. SHould we always use `calloc()`, as this gives us zero'd memory?